src/Symfony/Component/DependencyInjection/ContainerBuilder.php

```diff
@@ -1186,6 +1210,96 @@ public function getEnvCounters()
         return $this->envCounters;
     }
 
+    private function addOverriddenGetters($id, Definition $definition, \ReflectionClass $class, $salt)
+    {
+        if ($class->isFinal()) {
+            throw new RuntimeException(sprintf('Unable to configure getter injection for service "%s": class "%s" cannot be marked as final.', $id, $class->name));
+        }
+        $getters = '';
+        foreach ($definition->getOverriddenGetters() as $name => $returnValue) {
+            $r = self::getGetterReflector($class, $name, $id, $type);
+            $visibility = $r->isProtected() ? 'protected' : 'public';
+            $name = var_export($name, true);
+            $getters .= <<<EOF
+
+{$visibility} function {$r->name}(){$type} {
+    \$c = \$this->container{$salt};
+    \$b = \$c->getParameterBag();
+    \$v = \$this->values{$salt}[{$name}];
+
+    foreach (\$c->getServiceConditionals(\$v) as \$s) {
+        if (!\$c->has(\$s)) {
+            return parent::{$r->name}();
+        }
+    }
+
+    return \$c->resolveServices(\$b->unescapeValue(\$b->resolveValue(\$v)));
+}
+EOF;
+        }
+
+        return $getters;
+    }
+
+    public static function getGetterReflector(\ReflectionClass $class, $name, $id, &$type)
```

Comment:
`getGetterReflector` is used in two places (ContainerBuilder & PhpDumper), so I'd prefer sharing the implementation. Even short, it's not trivial.